cloud/docker_client.go

```diff
@@ -149,57 +152,99 @@ func (c *dockerClientImpl) EnsureImageDownloaded(ctx context.Context, h *host.Ho
 		// Image already exists
 		return imageName, nil
 	} else if strings.Contains(err.Error(), "No such image") {
-
-		// Extend http client timeout for ImageImport
-		normalTimeout := c.httpClient.Timeout
-		dockerClient, err = c.changeTimeout(h, imageImportTimeout)
-		if err != nil {
-			return "", errors.Wrap(err, "Error changing http client timeout")
+		settings := *h.Distro.ProviderSettings
+		if settings["build_type"] == distro.DockerImageBuildTypeImport {
+			grip.Info(message.Fields{
+				"operation": "EnsureImageDownloaded",
+				"details":   "start importing image",
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
+			err = c.importImage(ctx, h, imageName, url)
+			grip.Info(message.Fields{
+				"operation": "EnsureImageDownloaded",
+				"details":   "done importing image",
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
+			return imageName, errors.Wrap(err, "error importing image")
+		} else if settings["build_type"] == distro.DockerImageBuildTypePull {
+			grip.Info(message.Fields{
+				"operation": "EnsureImageDownloaded",
+				"details":   "start pulling image",
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
+			err = c.pullImage(ctx, h, url, settings["docker_registry_user"].(string), settings["docker_registry_pw"].(string))
+			grip.Info(message.Fields{
+				"operation": "EnsureImageDownloaded",
+				"details":   "done pulling image",
+				"duration":  time.Since(start),
+				"span":      time.Since(start).String(),
+			})
+			return imageName, errors.Wrap(err, "error pulling image")
 		}
+		return imageName, errors.Errorf("unrecognized image build method: %s", (*h.Distro.ProviderSettings)["build_type"])
```

Comment:
I feel like getting (*h.DistroProviderSettings)["build_id"] to be an error but not panic seems slim. maybe something narrower.

Reply:
oops, fixed